### DISCUSSION TITLE
Segreating xunit's testadapter .dll from unit test .dll files

### DISCUSSION THREAD

**Discussion by ByteMe666** [xunit/xunit#2725] [Question] [answered]
Hi,

I am running into the issue where [the solution is documented here](https://xunit.net/docs/getting-test-results-in-azure-devops).
However, I am not using Test selection/Test files, I am using Test Plan.

I am doing this in a Release Pipeline, which restricts me to UI, because I am trying [to automate test cases from Test Plans](https://learn.microsoft.com/en-us/azure/devops/test/run-automated-tests-from-test-hub?view=azure-devops).
Because the "Test Plan" option does not allow to append !**\xunit.runner.visualstudio.testadapter.dll anywhere, the test adapter is being picked up as a test assembly and I get a warning.

There is a menu to tell Azure DevOps to look for test assemblies in a specific folder, so my goal now is to put my test assemblies in a specific folder and segregate xunit.runner.visualstudio.testadapter.dll from them.

I am developing my xUnit tests as a standalone project standing on its own two feet, I created it in Visual Studio and let Visual Studio set up the project for me, I just have to click build and I see the .dll files appear. They are all put together in the same output folder. I looked at my .csproj file generated by Visual Studio: it is very short and nothing in it looks like instructions where to drop the artifacts after they've been built, so it doesn't look like it is defined there and that I can modify it there.

Where do I have to look ?

Thank you very much,
Arno

**Comment by bradwilson** [comment, marked as answer]
The naming of the test adapter is not negotiable; it's a requirement of VSTest.

> Nuget based Adapters in VS IDE. For seamless integration with adapters shipping as nuget packages, in VS IDE we automatically pick up referenced nuget packages that have files which satisfy *.TestAdapter.dll in their name and provide these files as adapters to the test platform.

https://github.com/microsoft/vstest/blob/main/docs/RFCs/0004-Adapter-Extensibility.md#specifying-an-adapter

The lack of a feature in Azure Pipelines is something that will have to be taken up with the Pipelines team.

**Reply by ByteMe666** [reply]
Hello !
Thank you for your answer. The naming of the adapter is not negotiable: ok ! But can it be put somewhere else than the same directory where the .dll test assemblies are located ?
Folder 1: Test assemblies
Folder 2: Test adapters
Thank you very much for taking time to reply !

**Reply by bradwilson** [reply]
The answer, unfortunately, is no.

There are essentially two different systems in play:

1. Test Explorer, in Visual Studio, knows about NuGet Package references and can peek inside the package to find the test adapter. Where it ends up living on disk is irrelevant.

2. The command line (things like `dotnet test`, `dotnet vstest`, `vstest.console.exe`, etc.) don't know anything about NuGet Package references. The place they look for the test adapters by default is in the same folder as the binary you're trying to test. If the test adapter file isn't on disk, the test can't run.

You can "override" the second condition with command line switches, like `/testadapterpath`, but that's not the default behavior. If we didn't put the file on disk with your test assembly, then nothing would work unless you passed this switch, and you'd have to know that the place it is in disk is in your home folder, like this:

![image](https://github.com/xunit/xunit/assets/16944/e8c553b2-bf24-4313-8f65-343f33c19253)

Except not only does the path change because the version changes, it might also change because the minimum supported platform version has changed:

![image](https://github.com/xunit/xunit/assets/16944/12b8945b-8065-4c8d-8b9a-e1967abe4fa1)

Realistically speaking, this isn't an option.

So, I go back to the point that Azure Pipelines should _already know_ that it should be excluding test adapters, because this is the way it's always been (since long before Azure Pipelines was even a thing). This is up to them to fix.
